README.md
Classic ray-tracing demo with checkerboard plane and a mirror sphere. In Clojure.

![https://pbs.twimg.com/media/DkGc73EXoAIUetw.jpg:large](https://pbs.twimg.com/media/DkGc73EXoAIUetw.jpg:large)


jayray/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-08-10
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-08-10
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/jayray/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/jayray/compare/0.1.0...0.1.1


jayray/test/jayray/core_test.clj
(ns jayray.core-test
  (:require [clojure.test :refer :all]
            [jayray.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


gen-pic.clj
(require '[clojure.string :as str])
(require '[clojure.java.io :as jio])

(load-file "quat.clj")

(def IW 1024)
(def IH 798)
(def oc [(+ 0.5 (/ IW -2)) (+ 0.5 (/ IH -2)) (/ IW 2)])
(def tilt 0.6)

(def sky-color [63 126 226])
(def plane-white [255 255 255])
(def plane-black [0 0 0])
(def error-color [255 0 0])

(def oo [0 -1.5 -0.6])
(def rr 0.6)

(declare camera-model find-color hit-sphere? hit-plane? touch-sphere? trace-reflection plane-texture)

(def pix
  (for [j (range IH)
        k (range IW)]
    (camera-model j k)))

(defn camera-model [j k]
  (let [cw [0 0 -2.0]
        qq [(- 1 (* tilt tilt)) (* tilt tilt) 0 0]
        vv (rotate qq (vsum [k j 0] oc))]
    (find-color cw vv 1)))

(defn find-color [cw vv iter]
  (or (hit-sphere? cw vv iter)
      (hit-plane? cw vv)
      sky-color))

(defn hit-sphere? [cw vv iter]
  (let [dd (touch-sphere? cw vv)]
    (if dd
      (trace-reflection cw dd vv iter)
      nil)))

(defn trace-reflection [cw dd vv iter]
  (let [newcw (vsum cw (vscale dd (normalze vv)))
        nn (vsub newcw oo)]
    (if (< 0 iter)
        (find-color newcw (reflect vv nn) (- iter 1))
        error-color)))

(defn hit-plane? [cw vv]
  (let [t (/ (* -1 (get cw 2)) (get vv 2))]
    (plane-texture cw vv t)))

(defn plane-texture [cw vv t]
  (if (< t 0)
    nil
    (let [px (+ (get cw 0) (* t (get vv 0)))
          py (+ (get cw 1) (* t (get vv 1)))]
       (if (> (norm [px py 0]) 50)
         nil
         (if (or (and (< (mod px 1.0) 0.5) (< (mod py 1.0) 0.5)) 
                 (and (> (mod px 1.0) 0.5) (> (mod py 1.0) 0.5)))
             (map int (vscale (/ 0.001 (* t t)) plane-white))
             (map int (vscale (/ 0.001 (* t t)) plane-black)))))))

(defn touch-sphere? [cw vv]
  (let [ll (normalze vv)
        oc (vsub cw oo)
        bb (dot ll oc)
        cc (- (dot oc oc) (* rr rr))
        delt (- (* bb bb) cc)    
        dd (- (* -1 bb) (Math/sqrt delt))]
    (if (and (> delt 0)
             (> 0 (+ (get cw 2) (* (get ll 2) dd)))
             (< 0 dd))
      dd
      nil)))

;; (def pix
;;   (for [j (range IH)
;;         k (range IW)]
;;     (let [r
;;           (+ 1e-7 (* 0.2 (Math/sqrt (+ (* k k) (* j j)))))]
;;       (int (* 255  (+ 0.5 ( * 0.5 (/ (Math/sin ( * (* 0.5 Math/PI) r))
;;                                      (Math/sqrt r)))))))))
    
;; output netbpm file
(defn write-to-file [file-name contents]
  (with-open [wtr (jio/writer file-name)]
    (dorun
     (.write wtr contents))))

(defn joined-values [vv]
  (str/join " "  vv))

(def p6-header ["P3" (str IW " " IH) "255"])

(def final-file-rows
  (conj p6-header
        (joined-values (flatten pix))
        ""))

(def final-file-contents
  (str/join "\n" final-file-rows))

(write-to-file "out.pnm" final-file-contents)


jayray/src/jayray/quat.clj
(ns jayray.quat
  (:gen-class))

(declare norm normalze dot vscale vsum vsub)

(defn qrot [a b c d]

  ;; [[(a * a + b * b - c * c - d * d), (2 * b * c - 2 * a * d), (2 * b * d + 2 * a * c)],
  ;;  [(2 * b * c + 2 * a * d), (a * a - b * b + c * c - d * d), (2 * c * d - 2 * a * b)],
  ;;  [(2 * b * d - 2 * a * c), (2 * c * d + 2 * a * b), (a * a - b * b - c * c + d * d)]]

  (let [aa (* a a)        ab (* a b)        ac (* a c)        ad (* a d)        bb (* b b)
        bc (* b c)        bd (* b d)        cc (* c c)        cd (* c d)        dd (* d d)]    
    [[(- (+ aa bb) cc dd),      (* 2 (- bc ad)),      (* 2 (+ bd ac))],     
     [(* 2 (+ bc ad)),      (- (+ aa cc) bb dd),      (* 2 (- cd ab))],
     [(* 2 (- bd ac)),      (* 2 (+ cd ab)),      (- (+ aa dd) bb cc)]]))
    
(defn rotate [qq vec]
  (let [qr (apply qrot qq)
        qx (get qr 0)
        qy (get qr 1)
        qz (get qr 2)]
    [(dot qx vec)
     (dot qy vec)
     (dot qz vec)]))
  
(defn reflect[dd nn]
  (vsub dd (vscale (* 2 (dot dd nn))
                   (normalze nn))))

(defn normalze [vv]
  (vec (map #(/ % (norm vv)) vv)))       

(defn norm [va]
  (Math/sqrt (dot va va)))

(defn dot [va vb]
  (+ (* (get va 0) (get vb 0))
     (* (get va 1) (get vb 1))
     (* (get va 2) (get vb 2))))

(defn vsum [va vb]
  [(+ (get va 0) (get vb 0))
   (+ (get va 1) (get vb 1))
   (+ (get va 2) (get vb 2))])

(defn vsub [va vb]
  [(- (get va 0) (get vb 0))
   (- (get va 1) (get vb 1))
   (- (get va 2) (get vb 2))])

(defn vscale [s vv]
  (vec (map #(* s %) vv)))
    
(def myq [(- 1 (* 0.1 0.1))          
          (* 0.1 0.1)
          0 0])
  
(def myv [0.0 1.0 0.0])

;; (println myq)

;; (println (apply qrot myq))

;; (println (rotate myq myv))


jayray/src/jayray/genpic.clj
(ns jayray.genpic
  (:gen-class))

(require '[clojure.string :as str])
(require '[clojure.java.io :as jio])

;; (load-file "quat.clj")
(use 'jayray.quat)

(def IW 1024)
(def IH 798)
(def oc [(+ 0.5 (/ IW -2)) (+ 0.5 (/ IH -2)) (/ IW 2)])
(def tilt 0.6)

(def sky-color [63 126 226])
(def plane-white [255 255 255])
(def plane-black [0 0 0])
(def error-color [255 0 0])

(def oo1 [0 -1.5 -0.6])
(def oo2 [1.2 -1.5 -0.6])
(def oo3 [-1.2 -1.5 -0.6])
(def rr 0.6)

(declare find-pixel-color find-ray-color hit-sphere? hit-plane? touch-sphere? trace-reflection
         plane-texture smallest-head? hit-sky vec-from-pixel make-ray test-ray)

(def pix
  (for [j (range IH)
        k (range IW)]
    (find-pixel-color [j k])))

(defn find-pixel-color [pix]
  (find-ray-color (vec-from-pixel pix)))

(defn vec-from-pixel [jk]
  (let [cw [0 0 -2.0]
        qq [(- 1 (* tilt tilt)) (* tilt tilt) 0 0]
        vv (rotate qq (vsum [(get jk 1) (get jk 0) 0] oc))]
    (make-ray cw vv)))

(defn make-ray [cw vv] {:orig cw :dir vv})

(defn find-ray-color [ray]
  (let [tests (test-ray ray)
        valid-tests (filter identity tests)
        the-hit (smallest-head? valid-tests)]
    (if (nil? the-hit)
      sky-color
      (the-hit))))

(defn test-ray [ray]
  [(hit-sphere? ray oo1)
   (hit-sphere? ray oo2)
   (hit-sphere? ray oo3)
   (hit-plane? ray)])


(defn hit-sky [] sky-color)

(defn smallest-head? [pairs]
  (if (empty? pairs)
    nil
    (:func (apply min-key (cons :dist pairs)))))

(defn hit-plane? [ray]
  (let [t (/ (* -1 (get (:orig ray) 2))
             (get (:dir ray) 2))
        yaya (plane-texture ray t)]
    (if yaya
      {:dist 1000 :func (fn [] yaya)}
      nil)))

(defn plane-texture [ray t]
  (if (< t 0)
    nil
    (let [px (+ (get (:orig ray) 0) (* t (get (:dir ray) 0)))
          py (+ (get (:orig ray) 1) (* t (get (:dir ray) 1)))]
       (if (> (norm [px py 0]) 50)
         nil
         (if (or (and (< (mod px 0.4) 0.2) (< (mod py 0.4) 0.2))
                 (and (> (mod px 0.4) 0.2) (> (mod py 0.4) 0.2)))
             (map int (vscale (/ 0.001 (* t t)) plane-white))
             (map int (vscale (/ 0.001 (* t t)) plane-black)))))))

(defn hit-sphere? [ray oo]
  (let [dd (touch-sphere? ray oo)]
    (if dd
      ;; [dd (fn [] (trace-reflection cw dd vv oo))]
      {:dist dd :func (fn [] (trace-reflection ray dd oo))}
      nil)))

(defn trace-reflection [ray dd oo]
  (let [newcw (vsum (:orig ray) (vscale dd (normalze (:dir ray))))
        nn (vsub newcw oo)]
    (find-ray-color (make-ray newcw (reflect (:dir ray) nn)))))

(defn touch-sphere? [ray oo]
  (let [ll (normalze (:dir ray))
        oc (vsub (:orig ray) oo)
        bb (dot ll oc)
        cc (- (dot oc oc) (* rr rr))
        delt (- (* bb bb) cc)
        dd (- (* -1 bb) (Math/sqrt delt))]
    (if (and (> delt 0)
             (> 0 (+ (get (:orig ray) 2) (* (get ll 2) dd)))
             (< 0.01 dd))
      dd
      nil)))

;;
;; output netbpm file
(defn write-to-file [file-name contents]
  (with-open [wtr (jio/writer file-name)]
    (dorun
     (.write wtr contents))))

(defn joined-values [vv]
  (str/join " "  vv))

(def p6-header ["P3" (str IW " " IH) "255"])

(def final-file-rows
  (conj p6-header
        (joined-values (flatten pix))
        ""))

(def final-file-contents
  (str/join "\n" final-file-rows))

; (write-to-file "out.pnm" final-file-contents)
